Fix SSE4.1-related issues

hebasto opened the pull request:
1. Fix the test for SSE4.1 intrinsics during build system configuration, which currently can be false positive, for example, when `CXXFLAGS="-mno-sse4.1"` provided.

	This PR fixes the test by adding the `_mm_blend_epi16` SSE4.1 function used in our codebase.

2. Guard `sha_x86_shani.cpp` code with `ENABLE_SSE41` macro as it uses the `_mm_blend_epi16` function from
the SSE4.1 instruction set.

	It is possible that SHA-NI is enabled even when SSE4.1 is disabled, which causes compile errors in the master branch.

Closes https://github.com/bitcoin/bitcoin/issues/28864.

laanwj commented:
Concept ACK

dergoegge commented:
Cool! Concept ACK

fanquake commented:
Concept ACK

luke-jr commented:
Weak Concept NACK: We should probably still build the SSE4.1 code if the compiler supports it. The flag should only affect the rest of the codebase, so that the software will run on a system without SSE4.1 (but still use the optimised crypto on systems that have it).

hebasto commented:
@luke-jr 
> Weak Concept NACK: We should probably still build the SSE4.1 code if the compiler supports it.

This PR branch still builds the SSE4.1 code if the compiler supports it, no?

hebasto committed:
build: Fix test for SSE4.1 intrinsics

This change uses the `_mm_blend_epi16` SSE4.1 function used in our code
and fixes false-positive cases, for example, when CXXFLAGS="-mno-sse4.1"
provided.

hebasto committed:
crypto: Guard code with `ENABLE_SSE41` macro

The code in `sha_x86_shani.cpp` uses the `_mm_blend_epi16` function from
the SSE4.1 instruction set. However, it is possible that SHA-NI is
enabled even when SSE4.1 is disabled.

This changes avoid compilation errors in such a condition.

hebasto commented:
Rebased.

fanquake commented:
I think this could be rebased again, given the recent build system changes, and removal of the sse4 lib.

hebasto commented:
> I think this could be rebased again, given the recent build system changes, and removal of the sse4 lib.

Sure! Rebased.

fanquake commented:
@theuni probably interested here, given you might als be refactoring the define usage?

willcl-ark commented:
Concept ACK.

I was tacking this today (missed this PR as GH search for "sse" yielded no results 🤦🏼‍♂️ )

First I tried re-implementing @luke-jr previous approach, but refactoring into macros: https://github.com/bitcoin/bitcoin/commit/cdaa9bfd6e63dc675cdd27e6d586f69423ee1748

Whilst this works well-enough, it didn't particularly fix https://github.com/bitcoin/bitcoin/issues/28864 in my opinion, so I tried https://github.com/bitcoin/bitcoin/compare/master...willcl-ark:bitcoin:detect-sse4.1

Now that I've found this PR, the mechanics of the two change-sets appear very similar, but @hebasto gates SHANI from the Makefile whereas I opted to gate in configure.ac. Not sure if either has any advantages, but I'm concept ACK here!

willcl-ark reviewed (approved):
tACK d440f13db02c82c842000abe4fe4d0c721a4ad3b

These changes fix the issues correctly in my testing on Ubuntu.

I do slightly prefer how my changeset printed a configure-time [warning message](https://github.com/bitcoin/bitcoin/commit/20b1fe92c87a86c8df4f1c7a0529a7d9e6710571#diff-49473dca262eeab3b4a43002adb08b4db31020d190caaad1594b47f1d5daa810R537) for the user that SHANI support would be disabled if SSE4.1 was not enabled, but overall prefer the approach of this fix.

theuni reviewed (approved):
utACK d440f13db02c82c842000abe4fe4d0c721a4ad3b

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/268.
